feat(avm)!: gt subtrace

MirandaWood opened the pull request:
Introduces the `gt` subtrace, a shared interface for `gt` and `ff_gt`, and redirects the ALU's LT/LTE ops to `gt` lookups.

(tests are a bit of a mess, sorry, I will make a follow up next week which brings ALU tests in line with eachother if the team wants!)

IlyasRidhuan committed:
feat(avm)! : comparison subtrace

MirandaWood committed:
feat: integrate gt circuit logic

MirandaWood committed:
chore: fix test. im stupid

MirandaWood committed:
feat: redirect LT/LTE to GT subtrace

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -45,21 +44,18 @@ MemoryValue Alu::lt(const MemoryValue& a, const MemoryValue& b)
     if (a.get_tag() == ValueTag::FF) {
-        // Emit the ff check event required (see lookup FF_LT) - note that we check b > a:
-        field_gt.ff_gt(b, a);
+        greater_than.gt(b_ff, a_ff);
     } else {
-        // We have excluded the field case => safe to downcast here for the max 128 bit range check:
-        lt_abs_diff = res ? static_cast<uint128_t>(b_ff - a_ff) - 1 : static_cast<uint128_t>(a_ff - b_ff);
+        greater_than.gt(static_cast<uint128_t>(b_ff), static_cast<uint128_t>(a_ff));
     }
```
Idea: you could have a `gt(MemoryValue, memoryValue)` method in the gadget, and in there just check the tag and call one or the other. Then here you could possibly just call the gadget and use the resulting boolean, without ever checking the tag.

Just something to consider, maybe you want to gather other opinions :)

MirandaWood replied:
Have added this! Lmk if looks ok (also @IlyasRidhuan)

fcarreiro commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -1,8 +1,8 @@
+include "gt.pil";
 include "precomputed.pil";
-include "range_check.pil";
 namespace alu;
```
sorry for the nit: could we have a newline between the namespace and the includes?

MirandaWood committed:
feat: overload gt, address comment

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.test.cpp:
```diff
@@ -80,14 +76,13 @@ TEST(AvmSimulationAluTest, NegativeAddTag)
     auto b = MemoryValue::from<uint32_t>(2);
 
-    EXPECT_CALL(range_check, assert_range(0, /*num_bits=*/32));
+    EXPECT_CALL(gt, gt(b, a));
```
Since we now overload such that:
`gt(MemoryValue, MemoryValue)` --> `gt(uint128, uint128)`,
can/should we test the nested call gets to the latter ok? i.e. something like `EXPECT_CALL(gt, gt(uint128(b), uint128(a)));`

fcarreiro replied:
I think if you are writing it like this (with the memoryvalues) the expect would be on the method with the memoryvalues. Since you are mocking, the inner redirection is really not something we'd need to test here.

You could test in the concrete GT object that
```
gt(MemoryValue, MemoryValue) --> gt(uint128, uint128),
```
However, this is easier in TS than in C++, since you have to mock either everything or nothing...

TL;DR: I guess from a behavioural PoV, this is enough for this test; and a test of the concrete GT class would have to test the 3 overloads separately without assuming they call each other (in theory).

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/simulation/gt.cpp:
```diff
@@ -0,0 +1,34 @@
+    });
+    return res;
+}
+bool GreaterThan::gt(const MemoryValue& a, const MemoryValue& b)
+{
+    FF a_ff = a.as_ff();
+    FF b_ff = b.as_ff();
+    // Note: not checking a_tag == b_tag here, as I think this is the job of
+    // memory-aware methods e.g. the ALU
+    if (a.get_tag() == ValueTag::FF) {
+        return gt(a_ff, b_ff);
+    }
+    return gt(static_cast<uint128_t>(a_ff), static_cast<uint128_t>(b_ff));
+}
```
I feel like this is the wrong place to do any checks on tags, but also we may want to guard against errors if `b_tag` is `FF` but a_tag is not?

IlyasRidhuan commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -147,43 +149,33 @@ pol commit sel_ff_lt_ops;
-sel_op_lt * (lt_ops_input_b - ib) + sel_op_lte * (lt_ops_input_b - ia) = 0;
+sel_op_lt * (lt_ops_input_b - ia) + sel_op_lte * (lt_ops_input_b - ib) = 0;
 
 #[LT_OPS_NEGATE_RESULT_C]
```
I guess to be clearer this needs to be `LTE..`

IlyasRidhuan commented:
I've stacked a [small PR](https://github.com/AztecProtocol/aztec-packages/pull/15684/files#diff-fa122fea8c3dfc3e7748edcbd4f380bdb6665e0cb3108ad977622b3fb9d61085) over this to fix up some stuff to help it to compile so i can use it in an upstream PR.

Feel free to merge when you're back and working on the tests

jeanmon reviewed (changes requested):
Great work overall. Please take a look at my comments.

Commit:
fix(avm)!: nits for Gt subtrace (#15684)

Please read [contributing guidelines](CONTRIBUTING.md) and remove this
line.

For audit-related pull requests, please use the [audit PR
template](?expand=1&template=audit.md).

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/simulation/gt.cpp:
```diff
@@ -0,0 +1,34 @@
+    FF b_ff = b.as_ff();
+    // Note: not checking a_tag == b_tag here, as I think this is the job of
+    // memory-aware methods e.g. the ALU
+    if (a.get_tag() == ValueTag::FF) {
```
Makes sense - that's around what I thought hence the no tag checking here! I'll just change the comment in my next commit, thanks

MirandaWood committed:
feat: address comments

jeanmon commented on barretenberg/cpp/pil/vm2/gt.pil:
```diff
@@ -0,0 +1,24 @@
+include "range_check.pil";
+
+namespace gt;
```
Deduplication is a good point. @MirandaWood Maybe you can just write a note as a potential future optimizations to consider. It is hard at this stage to know whether the deduplication saves more time than minimizing the extra-range checks.

MirandaWood committed:
chore: add comment
